### PR TITLE
Add events for STATUS packets.

### PR DESCRIPTION
**Pull request by thiccaxe**
Added StatusRequestEvent to edit or cancel the StatusRequestPacket
Added HandshakeEvent to determine protocol version and other data in HandshakePacket

### CODE REVIEW

**Comment by kezz** [comment]
Looks good to me! We could think about deprecating the `init` method that supplies the response data as an event makes a lot more sense.

File: src/main/java/net/minestom/server/event/server/StatusRequestEvent.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import org.jetbrains.annotations.NotNull;
+
+
+/**
+ * Called by a StatusRequestPacket.
+ * Can be used to modify the {@link ResponseData} response data.
+ */
+public class StatusRequestEvent extends Event implements CancellableEvent {
+
+    private final ResponseData responseData;
```
Comment (mworzala) [inline src/main/java/net/minestom/server/event/server/StatusRequestEvent.java:16]: Is there any reason to not just add the response data fields here? Otherwise every use case is just
```java
ResponseData data = event.getResponseData();
// ...
```

**Review by mworzala — COMMENTED** [review]
Looks good to me. I agree with Kezz on deprecating the init method.

**Comment by thiccaxe** [comment]
Not sure if I deprecated the ResponseDataConsumer correctly in 2dec361, please lmk.

File: src/main/java/net/minestom/server/event/server/StatusRequestEvent.java [hunk 2]
```diff
@@ -58,4 +63,120 @@ public boolean isCancelled() {
...
+
+    // Shortcut Methods
+    /**
+     * Sets the name for the response.
+     *
+     * @param name The name for the response data.
+     * @deprecated Use {@link #setVersion(String)}
+     */
+    @Deprecated
+    public void setName(String name) {
```
Comment (kezz) [inline src/main/java/net/minestom/server/event/server/StatusRequestEvent.java:75]: As this is new API we don't need these methods.

File: src/main/java/net/minestom/server/event/server/StatusRequestEvent.java [hunk 3]
```diff
@@ -58,4 +63,120 @@ public boolean isCancelled() {
...
+    }
+
+    /**
+     * Adds a player to the response.
+     *
+     * @param name The name of the player.
+     * @param uuid The unique identifier of the player.
+     */
+    public void addPlayer(String name, UUID uuid) {
+    }
```
Comment (kezz) [inline src/main/java/net/minestom/server/event/server/StatusRequestEvent.java:142]: Seems to be missing some code?

File: src/main/java/net/minestom/server/event/server/StatusRequestEvent.java [hunk 4]
```diff
@@ -58,4 +63,120 @@ public boolean isCancelled() {
...
+
+    /**
+     * Sets the response description.
+     *
+     * @param description The description for the response data.
+     * @deprecated Use {@link #setDescription(Component)}
+     */
+    @Deprecated
+    public void setDescription(String description) {
```
Comment (kezz) [inline src/main/java/net/minestom/server/event/server/StatusRequestEvent.java:159]: Same here, I wouldn't bother with carrying these deprecated methods over.

File: src/main/java/net/minestom/server/network/packet/client/handshake/HandshakePacket.java [hunk 5]
```diff
@@ -81,20 +82,21 @@ public void process(@NotNull PlayerConnection connection) {
...
-                // Happen when a client ping the server, ignore
+                // Happen when a client ping the server, trigger HandshakeEvent
+                MinecraftServer.getGlobalEventHandler().callEvent(HandshakeEvent.class, new HandshakeEvent(null, serverPort, protocolVersion, connection));
                 return;
             }
         }
 
         switch (nextState) {
             case 1:
+                MinecraftServer.getGlobalEventHandler().callEvent(HandshakeEvent.class, new HandshakeEvent(serverAddress, serverPort, protocolVersion, connection));
```
Comment (thiccaxe) [inline src/main/java/net/minestom/server/network/packet/client/handshake/HandshakePacket.java:99]: Should this line go before or after the ConnectionState is set to STATUS?

**Review by mworzala — APPROVED** [review]

**Comment by TheMode** [comment]
Is the name `HandshakeEvent`really appropriate? As the handshake packet is for state change, not necessarily ping

**Comment by thiccaxe** [comment]
> Is the name `HandshakeEvent`really appropriate? As the handshake packet is for state change, not necessarily ping

Maybe StasusHandshakeEvent (as it is a handshake packet with next state set to Stasus)

In the future a LoginHandshakeEvent could possibly be added.

Alternatively, call the event for both stasus changes and have that be a field in the class

**Comment by TheMode** [comment]
Bukkit has `ServerListPingEvent` not sure we need to find something different, need more opinion

**Comment by kezz** [comment]
I think `ServerListPingEvent` is probably a better description of what it does - although it *is* a handshake packet, it could be named so it's easier to know exactly what it does and what triggers the event.

**Comment by thiccaxe** [comment]
As it is implemented in Paper, it would require storing the data from the handshake packet in PlayerConnection or otherwise transferring the data between the packets.

The paper event has [this](https://papermc.io/javadocs/paper/1.16/com/destroystokyo/paper/network/StatusClient.html) class to get store the handshake packet's data.

The bukkit implementation also works, although it is missing this extra data

File: src/main/java/net/minestom/server/MinecraftServer.java [hunk 6]
```diff
@@ -742,7 +745,10 @@ public static void setShouldProcessNettyErrors(boolean processNettyErrors) {
      * @param port                 the server port
      * @param responseDataConsumer the response data consumer, can be null
      * @throws IllegalStateException if called before {@link #init()} or if the server is already running
+     *
+     * @deprecated use {@link #start(String, int)} and listen to the {@link net.minestom.server.event.server.StatusRequestEvent} event instead of ResponseDataConsumer
      */
+    @Deprecated
     public void start(@NotNull String address, int port, @Nullable ResponseDataConsumer responseDataConsumer) {
```
Comment (kezz) [inline src/main/java/net/minestom/server/MinecraftServer.java:752]: This method could overload the other one? Just set the consumer and then call the other init method.

File: src/main/java/net/minestom/server/event/server/ServerListPingEvent.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    public void setOnline(int online) {
+        responseData.setOnline(online);
+    }
+
+    /**
+     * Adds some players to the response.
+     *
+     * @param players the players
+     */
+    public void addPlayer(Iterable<Player> players) {
```
Comment (kezz) [inline src/main/java/net/minestom/server/event/server/ServerListPingEvent.java:115]: This should be @NotNull annotated

File: src/main/java/net/minestom/server/event/server/ServerListPingEvent.java [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        responseData.addPlayer(players);
+    }
+
+    /**
+     * Adds a player to the response.
+     *
+     * @param player the player
+     */
+    public void addPlayer(Player player) {
```
Comment (kezz) [inline src/main/java/net/minestom/server/event/server/ServerListPingEvent.java:124]: Also @NotNull annotated

File: src/main/java/net/minestom/server/event/server/ServerListPingEvent.java [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        addPlayer(player.getUsername(), player.getUuid());
+    }
+
+    /**
+     * Adds a player to the response.
+     *
+     * @param name The name of the player.
+     * @param uuid The unique identifier of the player.
+     */
+    public void addPlayer(String name, UUID uuid) {
```
Comment (kezz) [inline src/main/java/net/minestom/server/event/server/ServerListPingEvent.java:134]: Also @NotNull annotated - there's a few parameters in this class missing the annotations

File: src/main/java/net/minestom/server/event/server/ServerListPingEvent.java [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+        responseData.setFavicon(favicon);
+    }
+
+    /**
+     * Get the server address a client used to connect.
+     * may be null
+     * @return the server address
+     */
+    public @Nullable String getClientServerAddress() {
+        return Objects.requireNonNull(connection.getHandshakeData()).getServerAddress();
```
Comment (kezz) [inline src/main/java/net/minestom/server/event/server/ServerListPingEvent.java:180]: If the handshake is null, this will throw an exception - is that what we want here?

File: src/main/java/net/minestom/server/network/player/PlayerConnection.java [hunk 11]
```diff
@@ -170,6 +172,23 @@ public ConnectionState getConnectionState() {
         return connectionState;
     }
 
+    /**
+     * Gets the HandshakeData from the most recent HandshakePacket.
+     *
+     * @return the most recent HandshakeData
+     */
+    @Nullable
```
Comment (kezz) [inline src/main/java/net/minestom/server/network/player/PlayerConnection.java:180]: This annotation should be inline

File: src/main/java/net/minestom/server/network/player/PlayerConnection.java [hunk 12]
```diff
@@ -170,6 +172,23 @@ public ConnectionState getConnectionState() {
...
+    public HandshakeData getHandshakeData() {
+        return handshakeData;
+    }
+
+    /**
+     * Sets the HandshakeData - usually from a HandshakePacket
+     */
+    public void setHandshakeData(HandshakeData handshakeData) {
```
Comment (kezz) [inline src/main/java/net/minestom/server/network/player/PlayerConnection.java:188]: Also missing an annotation here

**Review by kezz — CHANGES_REQUESTED** [review]
Love the event - it's a lot more intuitive now but also exposing the handshake info in the event is pretty dang powerful

File: src/main/java/net/minestom/server/network/player/FakePlayerConnection.java [hunk 13]
```diff
@@ -41,4 +42,28 @@ public void setPlayer(Player player) {
         Check.argCondition(!(player instanceof FakePlayer), "FakePlayerController needs a FakePlayer object");
         super.setPlayer(player);
     }
+
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/player/FakePlayerConnection.java:45]: FakePlayer don't send any handshake packet, theres no need for code duplication

File: src/main/java/net/minestom/server/network/player/NettyPlayerConnection.java [hunk 14]
```diff
@@ -238,6 +240,24 @@ public void setRemoteAddress(@NotNull SocketAddress remoteAddress) {
...
+     */
+    public int getProtocolVersion() {
+        return protocolVersion;
+    }
+
+    /**
+     * Changes the protocol version of a connecting client.
+     * @param protocolVersion the protocol version of the client
+     */
+    public void setProtocolVersion(int protocolVersion) {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/player/NettyPlayerConnection.java:256]: See 'NettyPlayerConnection#refreshServerInformation'

File: src/main/java/net/minestom/server/network/player/NettyPlayerConnection.java [hunk 15]
```diff
@@ -276,11 +296,20 @@ public void UNSAFE_setLoginUsername(@NotNull String loginUsername) {
...
+    public @Nullable String getServerAddress() {
         return serverAddress;
     }
 
+    /**
+     * Set the server address the client used to connect.
+     *
+     * @param serverAddress the server address
+     */
+    public void setServerAddress(@Nullable String serverAddress) {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/player/NettyPlayerConnection.java:308]: Same as above

File: src/main/java/net/minestom/server/network/player/NettyPlayerConnection.java [hunk 16]
```diff
@@ -349,16 +387,6 @@ public void setConnectionState(@NotNull ConnectionState connectionState) {
         }
     }
 
-    /**
-     * Used in {@link net.minestom.server.network.packet.client.handshake.HandshakePacket} to change the internal fields.
-     *
-     * @param serverAddress the server address which the client used
-     * @param serverPort    the server port which the client used
-     */
-    public void refreshServerInformation(@Nullable String serverAddress, int serverPort) {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/player/NettyPlayerConnection.java:358]: Why did you remove it?

**Review by TheMode — CHANGES_REQUESTED** [review]

File: src/main/java/net/minestom/server/network/packet/client/handshake/HandshakePacket.java [hunk 17]
```diff
@@ -92,19 +90,18 @@ public void process(@NotNull PlayerConnection connection) {
...
             }
         }
-
         switch (nextState) {
             case 1:
                 connection.setConnectionState(ConnectionState.STATUS);
                 break;
             case 2:
+                if (connection instanceof NettyPlayerConnection) {
+                    ((NettyPlayerConnection) connection).setProtocolVersion(protocolVersion);
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/client/handshake/HandshakePacket.java:99]: You can put the protocol in `refreshServerInformation`, there is no reason to have a new setter

**Review by kezz — APPROVED** [review]
Barring the changes TheMode has mentioned, LGTM!

File: src/main/java/net/minestom/server/network/packet/client/status/StatusRequestPacket.java [hunk 18]
```diff
@@ -29,10 +30,17 @@ public void process(@NotNull PlayerConnection connection) {
         if (consumer != null)
             consumer.accept(connection, responseData);
 
-        ResponsePacket responsePacket = new ResponsePacket();
-        responsePacket.jsonResponse = responseData.build().toString();
+        // Call event
+        ServerListPingEvent statusRequestEvent = new ServerListPingEvent(responseData, connection);
+        MinecraftServer.getGlobalEventHandler().callEvent(ServerListPingEvent.class, statusRequestEvent);
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/client/status/StatusRequestPacket.java:35]: See `EventHandler#callCancellableEvent`

File: src/main/java/net/minestom/server/event/server/ServerListPingEvent.java [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+import net.minestom.server.event.Event;
+import net.minestom.server.network.player.NettyPlayerConnection;
+import net.minestom.server.network.player.PlayerConnection;
+import net.minestom.server.ping.ResponseData;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.UUID;
+
+public class ServerListPingEvent extends Event implements CancellableEvent {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/event/server/ServerListPingEvent.java:16]: I believe that only #getResponseData should be used to modify the ping result, otherwise we end up with a lot of code duplication and be hard to maintain

**Review by TheMode — CHANGES_REQUESTED** [review]

File: src/main/java/net/minestom/server/event/server/ServerListPingEvent.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
+package net.minestom.server.event.server;
+
+import net.minestom.server.event.CancellableEvent;
+import net.minestom.server.event.Event;
+import net.minestom.server.network.player.PlayerConnection;
+import net.minestom.server.ping.ResponseData;
+import org.jetbrains.annotations.NotNull;
+
+public class ServerListPingEvent extends Event implements CancellableEvent {
```
Comment (kezz) [inline src/main/java/net/minestom/server/event/server/ServerListPingEvent.java:9]: This should have some javadoc.

**Review by kezz — APPROVED** [review]
Just a minor comment but LGTM!

**Review by TheMode — APPROVED** [review]
